src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,533 @@
+        size_t idx = 0, length = 0;
+        int r = 0;
+
+        r = fetch_bpf_maps(&map_list);
+        if (r < 0)
+                return r;
+
+        length = json_variant_elements(map_list);
+        JSON_VARIANT_ARRAY_FOREACH(v, map_list) {
+                if (idx < length - 1) {
+                        r = varlink_notify(link, v);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to varlink notify for map: %m");
+                } else {
+                        /* varlink reply on final element only */
+                        r = varlink_reply(link, v);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to varlink reply for map: %m");
+                }
+                idx++;
+        }
+
+        return 0;
+}
+
+static int vl_method_describe_programs(Varlink *link, JsonVariant *parameters, VarlinkMethodFlags flags, void *userdata) {
+        JsonVariant *v = NULL, *prog_list = NULL;
+        size_t idx = 0, length = 0;
+        int r = 0;
+
+        r = fetch_bpf_programs(&prog_list);
+        if (r < 0)
+                return r;
+
+        length = json_variant_elements(prog_list);
+        JSON_VARIANT_ARRAY_FOREACH(v, prog_list) {
+                if (idx < length - 1) {
+                        r = varlink_notify(link, v);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to varlink notify for program: %m");
+                } else {
+                        /* varlink reply on final element only */
+                        r = varlink_reply(link, v);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to varlink reply for program: %m");
+                }
+                idx++;
+        }
+
+        return 0;
+}
+
+static int manager_bind_varlink(Manager *m) {
+        const char *socket_path;
+        int r;
+
+        assert(m);
+        assert(!m->varlink_server);
+
+        r = varlink_server_new(&m->varlink_server, VARLINK_SERVER_INHERIT_USERDATA);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate varlink server object: %m");
+
+        varlink_server_set_userdata(m->varlink_server, m);
+
+        r = varlink_server_bind_method_many(
+                        m->varlink_server,
+                        "io.systemd.Bpf.DescribeMaps", vl_method_describe_maps,
+                        "io.systemd.Bpf.DescribePrograms", vl_method_describe_programs
+        );
+        if (r < 0)
+                return log_error_errno(r, "Failed to register varlink methods: %m");
+
+        (void) mkdir_p("/run/systemd/bpf", 0755);
+        socket_path = "/run/systemd/bpf/io.systemd.Bpf";
+
+        r = varlink_server_listen_address(m->varlink_server, socket_path, 0600);
+        if (r < 0)
+                return log_error_errno(r, "Failed to bind to varlink socket: %m");
```

Comment:
Please include the path to the socket in the error message, that is the most important piece of information for a user.